### PR TITLE
Version-4 candidate

### PR DESCRIPTION
**Pull request by smyslov**
Mostly inspired by Paul Wouters' comments:

   o  Introduction and initial sectiona are reorganized.
   o  More clarifications for error handling added.
   o  ASCII arts displaying SA payload are added.
   o  Clarification for handling multiple round trips key exchange
      methods added.
   o  DoS concerns added into Security Considerations section.
   o  Explicitly allow scenario when additional key exchanges are
      performed only after peers are authenticated.

### CODE REVIEW

**Review by grahamietf — COMMENTED** [review]
I've been through this and it looks good.

I've fixed one minor typo.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
            <t> We assume that new Transform Type 4 identifiers will be assigned later to the
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
            to be of any type - either post-quantum or classical (EC)DH
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
            one.  This approach allows any combination of defined key exchange methods
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
            For hypothetical future key exchange methods requiring multiple round trips to complete, a separate document should define how such
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:521]: Is it better to have the Additional Key Exchange entries written as a list, rather than having them in a single line?

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Did you mean?
```suggestion
                responder selects this value for all of them, then no IKE_INTERMEDIATE exchanges will take place between the peers.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:570]:
```suggestion
                (note that they still may take place for other purposes).
```

Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:570]: I am a little lost here, but what did you mean by "they still may take place for other purposes"?

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:606]: I think we should avoid specific algorithm names (i.e. `FRODOKEM_976_AES`, `SIKE_P434` and `KYBER512`). They are still NIST candidates and we don't want to be seen to "favour" these algorithms, or influencing selection, etc. Perhaps we can use some abstract name, e.g.: NIST_PQ_KEM_1, NIST_PQ_KEM_2, etc.

Or if anyone can come up with a better name, happy to use that too.

Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:606]: The transform `AKE1` is not listed and I guess this is on-purpose?

Do you envisage that we will have a standard list of what transform IDs that will be included in `AKE1`,  `AKE2`, etc?

Just curious, in practice, what circumstances we will skip `AKE1`. Usually, the transform type `AKEi` would be preferred than `AKEj` where `i < j`.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:611]: Do we need to describe that both `AKE2` and `AKE3` are mandatory? If the peer doesn't support both of these, the connection cannot be established.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:646]: Surely we still have the proposal for Transform Type 4, right? Or did you mean that even both peers cannot agree on the transform ID for Transform Type 4?

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
                For the methods requiring multiple round trips, a separate document should define how such
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Perhaps this?

```suggestion
        key exchange methods, where many of them are more memory and/or CPU intensive than the classical counterparts.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
        In this case at the point when resource-intensive operations are required,
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
            <t>Introduction and initial sections are reorganized.</t>
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
                in these situations.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: I think this should be:
```suggestion
                additional key exchanges and thus never start the remaining IKE_FOLLOWUP_KE exchanges.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
        This is similar to the scenario described in <xref target="RFC8784" />.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Perhaps:

```suggestion
        Depending on peers' policy, this scenario may or may not be appropriate.
```

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]:
```suggestion
        <t> The authors would like to thank Frederic Detienne and Olivier
```

**Review by ctjhai — APPROVED** [review]
What do you think of adding an Appendix section, showing some message flows for various scenarios, e.g.:
- Responder selects NONE on all AKEs so the resulting exchanges don't use IKE_INTERMEDIATE
- Peers cannot agree on one or more mandatory AKEs, and therefore the negotiation results in NO_PROPOSAL_CHOSEN error
- No AKE is used in IKE_SA_INIT, but AKE is used in CREATE_CHILD_SA
- etc

I am happy to attempt this, if you are okay with this. Unless you already have these up in your sleeve :-)

**Comment by ctjhai** [comment]
Should we also add specific error messages e.g. INVALID_SYNTAX for some erroneous scenarios, like what I attempted to do in [my PR](https://github.com/post-quantum/ietf-pq-ikev2/pull/11)? I am happy to update [my PR](https://github.com/post-quantum/ietf-pq-ikev2/pull/11) again once this one is merged to `master` branch. That's if you think it is worth having them.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:521]: I don't know. So, I think that it's OK in current form, but I don't object if instead they are shown as a list.

**Comment by smyslov** [comment]
I tried to cover this situations, but I was preparing the text candidate in a hurry, so I probably overlooked something.
So, my idea is that we don't need to specify error handling in situations which are not specified in our base document - RFC 7296. For example, there is nothing in RFC 7296 about how to behave if DH Group number in KE payload received by initiator in response message doesn't match the value peers agreed to use. Obviously, it's a protocol error - what to do depends on implementation, most probably just delete IKE SA w/o even sending INVALID_SYNTAX (because of the RFC 7296's rule that errors in responces must not cause initiation of new exchanges).
So, I think we should carefully look through the all places when error handling is assumed and specify only this behaviour, which is really necessary (like sending NO_PROPOSAL_CHOSEN). I'll do another pass of review my and your PRs soon...

**Comment by smyslov** [comment]
> What do you think of adding an Appendix section, showing some message flows for various scenarios, e.g.:
> 
> * Responder selects NONE on all AKEs so the resulting exchanges don't use IKE_INTERMEDIATE
> * Peers cannot agree on one or more mandatory AKEs, and therefore the negotiation results in NO_PROPOSAL_CHOSEN error
> * No AKE is used in IKE_SA_INIT, but AKE is used in CREATE_CHILD_SA
> * etc
> 
> I am happy to attempt this, if you are okay with this. Unless you already have these up in your sleeve :-)

No, I don't :-) It'll be great if you do this. I believe three scenarios that you mentioned before are enough...

**Comment by ctjhai** [comment]
> No, I don't :-) It'll be great if you do this. I believe three scenarios that you mentioned before are enough...

Hi @smyslov, apologies for the delay in getting this, but the three scenarios are now available here: https://github.com/post-quantum/ietf-pq-ikev2/pull/13.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:570]: Well, IKE_INTERMEDIATE can be used not only for performing additional key exchanges, but for other purposes too (e.g. look at draft-smyslov-ipsecme-ikev2-auth-announce). So, if peers negotiate no AKEs, they can still perform IKE_INTERMEDIATE for some other purposes.
If you can rephrase this sentence to make it more clear, I'd be appreciated.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:606]: > I think we should avoid specific algorithm names (i.e. `FRODOKEM_976_AES`, `SIKE_P434` and `KYBER512`). They are still NIST candidates and we don't want to be seen to "favour" these algorithms, or influencing selection, etc. Perhaps we can use some abstract name, e.g.: NIST_PQ_KEM_1, NIST_PQ_KEM_2, etc.
> 
> Or if anyone can come up with a better name, happy to use that too.

OK, makes sense.

Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:606]: > The transform `AKE1` is not listed and I guess this is on-purpose?
> 
> Do you envisage that we will have a standard list of what transform IDs that will be included in `AKE1`, `AKE2`, etc?
> 
> Just curious, in practice, what circumstances we will skip `AKE1`. Usually, the transform type `AKEi` would be preferred than `AKEj` where `i < j`.

Yes, on purpose to illustrate that there is no requirement in the draft that proposed AKEs must start from AKE1, followed by AKE2 (if any) etc. It's perfectly valid to propose only, say, AKE3 and AKE6. I can imagine that some implementers may associate specific AKEs with specific PQ KEMs, e.g. AKE1 contains lattice-based KEMs, AKE2 - code based, AKE3 - isogeny based. In this case depending on local policy (say, security officer decided to only rely on code-based KEMs), initiator may only propose AKE2. I don't know how practical it is, just don't want to add restrictions to prevent this scenario.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:611]: Do you think it is not clear from the fact that neither of them contains NONE?

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (smyslov) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:646]: I meant that AKE Transforms types (not Transform IDs) must be treated as unknown if IKE_INTERMEDIATE wasn't negotiated. So, according to RFC 7296 responder must skip any proposals with unknown transforms. If it finds another proposal with no AKE transform, then it evaluates Transform IDs for every transform in this proposal. If in this proposal Transform Type 4 contains transform ID for PQ KEM, then it's OK, and peers will perform a single PQ Key Exchange in IKE_SA_INIT, with no AKEs.

**Comment by smyslov** [comment]
Sorry, I completely forgot to address your comments :-( Now done.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:606]: Ok, yep that makes sense and I have a follow-on question on this subject and it might be related to what Paul asked on the mailing list. But let's not hold this up, I'll follow up with you separately on this question.

I'll merge this PR and I'll rename some of the specific PQ algorithms to generic names, i.e. FRODOKEM_976_AES to PQ_KEM_X.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:646]: Ok, that makes sense.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:611]: I meant, explicitly say that if any of the mandatory AKEs cannot be negotiated, then the connection cannot be established. But let's leave it as it is for now.

**Pull request merged**
